expression/builtin_cast.go

```diff
@@ -1955,9 +1955,16 @@ func WrapWithCastAsString(ctx sessionctx.Context, expr Expression) Expression {
 	if exprTp.Tp == mysql.TypeNewDecimal && argLen != types.UnspecifiedFsp {
 		argLen += 3
 	}
+
 	if exprTp.EvalType() == types.ETInt {
 		argLen = mysql.MaxIntWidth
+		// For TypeBit, castAsString will make length as int(( bit_len + 7 ) / 8) bytes due to
+		// TiKV needs the bit's real len during calculating, eg: ascii(bit).
+		if exprTp.Tp == mysql.TypeBit {
+			argLen = (exprTp.Flen + 7) / 8
+		}
```

Comment:
Is there any testcase to cover this? eg: push ascii bit column down and assert the result.

Reply:
Ok, I will add it.